src/Raven.Client/Util/WeakReferencingTimer.cs

```diff
@@ -0,0 +1,64 @@
+﻿using System;
+using System.Threading;
+
+namespace Raven.Client.Util;
+
+internal class WeakReferencingTimer : IDisposable
+{
+    internal delegate void WeakReferencingTimerCallback(object state);
+
+    private readonly Timer _timer;
+
+    public WeakReferencingTimer(WeakReferencingTimerCallback callback, object state, TimeSpan dueTime, TimeSpan period)
+    {
+        if (callback == null)
+            throw new ArgumentNullException(nameof(callback));
+
+        var internalTimerState = new TimerState
+        {
+            State = state is not null ? new WeakReference<object>(state) : null,
+            Callback = new WeakReference<WeakReferencingTimerCallback>(callback)
+        };
+
+        _timer = new Timer(StaticCallback, internalTimerState, dueTime, period);
+
+        internalTimerState.Timer = _timer;
```

Comment:
this looks racy, there is a chance of NRE in the StaticCallback because of that. Wondering if we should not initialize it first, assign it and then apply timer.Change?

Reply:
You're correct. Fixed